### PR TITLE
319 - Emit Booking cancelled domain event

### PR DESCRIPTION
**Pull request by brownbearsoftware**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingService.kt [hunk 1]
```diff
@@ -884,6 +888,67 @@ class BookingService(
       )
     }
 
+    if (booking.service == ServiceName.approvedPremises.value && booking.application != null && user != null) {
+      val dateTime = OffsetDateTime.now()
+
+      val domainEventId = UUID.randomUUID()
+
+      val offenderDetails = when (val offenderDetailsResult = offenderService.getOffenderByCrn(booking.crn, user.deliusUsername, user.hasQualification(UserQualification.LAO))) {
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingService.kt:896]: Given this is going to an internal service, do we need to worry about if the user has the qualification here?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingService.kt [hunk 2]
```diff
@@ -884,6 +888,67 @@ class BookingService(
...
+        is AuthorisableActionResult.Success -> offenderDetailsResult.entity
+        is AuthorisableActionResult.Unauthorised -> throw RuntimeException("Unable to get Offender Details when creating Booking Cancelled Domain Event: Unauthorised")
+        is AuthorisableActionResult.NotFound -> throw RuntimeException("Unable to get Offender Details when creating Booking Cancelled Domain Event: Not Found")
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingService.kt:899]: I wonder if we should just return null here if we can't access Delius (and maybe log upstream?), as it's only fetching the NOMS number, which we can get at a later date. Doesn't make sense to block a user action for something which is not necessarily their fault.

**Review by pezholio — APPROVED** [review]
LGTM - just a couple of things to think about (which I think affects other domain events too)

**Review by danlivings-dxw — APPROVED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingService.kt (hunk 1 above)
Comment (brownbearsoftware) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingService.kt:896]: Good point, have updated this & added another commit to address for the existing events

**Pull request merged**